File: android/src/main/java/com/apppresser/onesignal/CapOneSignal.java
package com.apppresser.onesignal;

import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.os.Build;
import android.provider.Settings;

import androidx.core.app.ActivityCompat;
import androidx.core.app.NotificationManagerCompat;
import com.getcapacitor.Logger;
import com.getcapacitor.Plugin;
import com.onesignal.OneSignal;
import com.onesignal.debug.LogLevel;
import com.onesignal.OSNotificationClickEvent;
import com.onesignal.OSNotificationClickListener;
import org.json.JSONObject;
import org.json.JSONException;

import com.getcapacitor.PluginCall;
import com.getcapacitor.PermissionState;
import android.Manifest;
import androidx.core.app.ActivityCompat;
import java.util.Iterator;
import com.getcapacitor.JSObject;

public class CapOneSignal implements OSNotificationClickListener {

    private Plugin plugin;
    private Context context;

    /**
     * Initialize OneSignal with application context and appId.
     */
    public void initialize(Context ctx, String appID) {
        this.context = ctx;
        
        if (appID == null || appID.isEmpty()) {
            Logger.warn("CapOneSignal", "initialize: appID is null/empty");
            return;
        }

        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.LOLLIPOP) {
            Logger.warn("CapOneSignal", "OneSignal SDK v5 requires Android API 21 (Lollipop) or newer. Device is running API " + Build.VERSION.SDK_INT);
            return;
        }

        if (!"REL".equals(Build.VERSION.CODENAME)) {
            Logger.info("CapOneSignal", "Running on a pre-release version of Android: " + Build.VERSION.CODENAME + ". Enabling verbose logging.");
            OneSignal.getDebug().setLogLevel(LogLevel.VERBOSE);
        }

        OneSignal.initWithContext(ctx.getApplicationContext(), appID);
        
        // Register click listener
        OneSignal.getNotifications().addClickListener(this);
        
        Logger.info("CapOneSignal", "OneSignal initialized with appID: " + appID);
    }

    public void setPlugin(Plugin plugin) {
        this.plugin = plugin;
    }

    @Override
    public void onClick(OSNotificationClickEvent event) {
        Logger.info("CapOneSignal", "Notification clicked: " + event.getNotification().getNotificationId());
        
        try {
            JSONObject clickData = new JSONObject();
            
            // Notification ID
            clickData.put("notificationId", event.getNotification().getNotificationId());
            
            // Launch URL
            String launchURL = event.getNotification().getLaunchURL();
            if (launchURL != null) {
                clickData.put("launchURL", launchURL);
                Logger.info("CapOneSignal", "Launch URL: " + launchURL);
                
                // Open URL if valid
                try {
                    Intent intent = new Intent(Intent.ACTION_VIEW, Uri.parse(launchURL));
                    intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
                    context.startActivity(intent);
                } catch (Exception e) {
                    Logger.error("CapOneSignal", "Failed to open launch URL: " + e.getMessage());
                }
            }
            
            // Additional Data - THIS IS THE KEY PART for your custom data!
            JSONObject additionalData = event.getNotification().getAdditionalData();
            if (additionalData != null) {
                Logger.info("CapOneSignal", "Additional Data: " + additionalData.toString());
                clickData.put("additionalData", additionalData);
            } else {
                Logger.info("CapOneSignal", "No additionalData in notification");
            }
            
            // Notify JavaScript listeners
            if (plugin != null) {
                Logger.info("CapOneSignal", "Notifying JavaScript listeners of notification click");
                
                JSObject jsClickData;
                try {
                    jsClickData = JSObject.fromJSONObject(clickData);
                } catch (JSONException e) {
                    Logger.error("CapOneSignal", "Error converting JSONObject to JSObject: " + e.getMessage());
                    jsClickData = new JSObject(); // Fallback to empty
                }
                
                plugin.notifyListeners("notificationClicked", jsClickData, true);
            } else {
                Logger.error("CapOneSignal", "ERROR: plugin reference is null, cannot notify JavaScript");
            }
            
        } catch (JSONException e) {
            Logger.error("CapOneSignal", "Error building click data: " + e.getMessage());
        }
    }

    /**
     * Set OneSignal log level. Accepts: VERBOSE, DEBUG, INFO, WARN, WARNING, ERROR, NONE
     */
    public void setLogLevel(String level) {
        if (level == null || level.isEmpty()) {
            Logger.warn("CapOneSignal", "setLogLevel: level is null/empty");
            return;
        }
        String upper = level.toUpperCase();

        // In OneSignal v5, use the LogLevel enum from com.onesignal.debug
        LogLevel logLevel;
        switch (upper) {
            case "VERBOSE":
                logLevel = LogLevel.VERBOSE;
                break;
            case "DEBUG":
                logLevel = LogLevel.DEBUG;
                break;
            case "INFO":
                logLevel = LogLevel.INFO;
                break;
            case "WARN":
            case "WARNING":
                logLevel = LogLevel.WARN;
                break;
            case "ERROR":
                logLevel = LogLevel.ERROR;
                break;
            case "NONE":
                logLevel = LogLevel.NONE;
                break;
            default:
                Logger.warn("CapOneSignal", "Unknown log level: " + level + ", defaulting to INFO");
                logLevel = LogLevel.INFO;
        }

        // Set SDK log level using the Debug manager in v5
        try {
            OneSignal.getDebug().setLogLevel(logLevel);
            Logger.info("CapOneSignal", "OneSignal log level set to: " + upper);
        } catch (Exception e) {
            Logger.warn("CapOneSignal", "Failed to set OneSignal log level: " + e.getMessage());
        }
    }

    public void requestPermission(Context ctx, PluginCall call, boolean fallbackToSettings) {
        // This method should now ONLY handle the legacy path.
        // The Android 13+ logic is (and should be) in CapOneSignalPlugin.java.

        boolean granted = areNotificationsEnabled(ctx);
        Logger.info("CapOneSignal", "Legacy permission check: notifications enabled = " + granted);

        if (!granted && fallbackToSettings) {
            Logger.info("CapOneSignal", "Notifications are not enabled. Opening settings for user.");
            openNotificationSettings(ctx);
        }

        // Always resolve the call for this legacy path. The "status" will be determined
        // by the calling method in the plugin by re-checking areNotificationsEnabled.
        call.resolve();
    }

    /**
     * Return whether notifications are enabled for the app.
     */
    public boolean areNotificationsEnabled(Context ctx) {
        return NotificationManagerCompat.from(ctx).areNotificationsEnabled();
    }

    /**
     * Open the app notification settings so the user can enable notifications.
     */
    public void openNotificationSettings(Context ctx) {
        Intent intent;
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            intent = new Intent(Settings.ACTION_APP_NOTIFICATION_SETTINGS).putExtra(Settings.EXTRA_APP_PACKAGE, ctx.getPackageName());
        } else {
            intent = new Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS).setData(Uri.fromParts("package", ctx.getPackageName(), null));
        }
        intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        ctx.startActivity(intent);
    }

    /**
     * Set an external user id for OneSignal (v5 API: login).
     */
    public void setExternalUserId(String userID) {
        if (userID == null || userID.isEmpty()) {
            Logger.warn("CapOneSignal", "setExternalUserId: userID is null/empty");
            return;
        }
        // Use setExternalUserId() to associate the current subscription with an external id
        OneSignal.login(userID);
        Logger.info("CapOneSignal", "setExternalUserId (login): " + userID);
    }

    /**
     * Clear external user id (v5 API: logout).
     */
    public void clearExternalUserId() {
        OneSignal.logout();
        Logger.info("CapOneSignal", "clearExternalUserId (logout) called");
    }

    /**
     * Add multiple tags to the user.
     */
    public void addTags(java.util.Map<String, String> tags) {
        if (tags == null || tags.isEmpty()) {
            Logger.warn("CapOneSignal", "addTags: tags is null/empty");
            return;
        }
        OneSignal.getUser().addTags(tags);
        Logger.info("CapOneSignal", "addTags: " + tags.toString());
    }

    /**
     * Remove multiple tags from the user.
     */
    public void removeTags(java.util.Collection<String> tagKeys) {
        if (tagKeys == null || tagKeys.isEmpty()) {
            Logger.warn("CapOneSignal", "removeTags: tagKeys is null/empty");
            return;
        }
        OneSignal.getUser().removeTags(tagKeys);
        Logger.info("CapOneSignal", "removeTags: " + tagKeys.toString());
    }

    /**
     * Add a single tag to the user.
     */
    public void addTag(String key, String value) {
        if (key == null || key.isEmpty() || value == null || value.isEmpty()) {
            Logger.warn("CapOneSignal", "addTag: key or value is null/empty");
            return;
        }
        OneSignal.getUser().addTag(key, value);
        Logger.info("CapOneSignal", "addTag: " + key + " = " + value);
    }

    /**
     * Remove a single tag from the user.
     */
    public void removeTag(String key) {
        if (key == null || key.isEmpty()) {
            Logger.warn("CapOneSignal", "removeTag: key is null/empty");
            return;
        }
        OneSignal.getUser().removeTag(key);
        Logger.info("CapOneSignal", "removeTag: " + key);
    }

}

File: ios/Sources/CapOneSignalPlugin/CapOneSignal.swift
import Capacitor
import Foundation
import OneSignalFramework
import UIKit

@objc public class CapOneSignal: NSObject, OSNotificationClickListener {

    @objc public static let shared = CapOneSignal()

    // Store weak reference to plugin for notifying JavaScript
    weak var plugin: CAPPlugin?

    public override init() { super.init() }

    @objc public func setPlugin(_ plugin: CAPPlugin) {
        self.plugin = plugin
    }

    @objc public func initialize(_ appID: String, plugin: CAPPlugin? = nil) {
        print("Initializing OneSignal with App ID: \(appID)")
        OneSignal.initialize(appID)

        // Set plugin reference if provided
        if let plugin = plugin {
            self.plugin = plugin
            print("CapOneSignal: Plugin reference set")
        }

        OneSignal.Notifications.addClickListener(CapOneSignal.shared)
    }

    @objc public func handleLaunchOptions(_ launchOptions: [UIApplication.LaunchOptionsKey: Any]?) {
        if let remoteNotif = launchOptions?[.remoteNotification] as? [String: Any] {
            print("DEBUG: App launched via Notification Tap!")

            // Build click data object for cold start
            var clickData: [String: Any] = [:]

            // Extract notification ID
            if let notifId = remoteNotif["id"] as? String {
                clickData["notificationId"] = notifId
            }

            if let custom = remoteNotif["custom"] as? [String: Any] {
                // Extract launch URL from 'custom' -> 'u'
                if let launchURL = custom["u"] as? String {
                    clickData["launchURL"] = launchURL

                    print("DEBUG: Found Launch URL in Cold Start: \(launchURL)")

                    // Open URL after small delay
                    DispatchQueue.main.asyncAfter(deadline: .now() + 0.5) {
                        if let url = URL(string: launchURL) {
                            UIApplication.shared.open(url, options: [:], completionHandler: nil)
                        }
                    }
                }

                // Extract additionalData from 'custom' -> 'a'
                if let additionalDataAny = custom["a"] as? [AnyHashable: Any] {
                    var convertedData: [String: Any] = [:]
                    for (key, value) in additionalDataAny {
                        if let stringKey = key as? String {
                            convertedData[stringKey] = value
                        }
                    }
                    clickData["additionalData"] = convertedData
                    print("DEBUG: Found Additional Data in Cold Start: \(convertedData)")
                }
            }

            // Notify JavaScript listeners for cold start clicks
            if let plugin = self.plugin {
                print("DEBUG: Notifying JavaScript listeners of cold start click")
                plugin.notifyListeners("notificationClicked", data: clickData, retainUntilConsumed: true)
            }
        }
    }

    @objc public func onClick(event: OSNotificationClickEvent) {
        print("OneSignal: Notification clicked!")

        // Build click data object
        var clickData: [String: Any] = [:]

        // Notification ID
        clickData["notificationId"] = event.notification.notificationId
        print("OneSignal: Notification ID: \(event.notification.notificationId)")

        // Launch URL
        if let launchURL = event.notification.launchURL {
            clickData["launchURL"] = launchURL
            print("OneSignal: Launch URL: \(launchURL)")

            DispatchQueue.main.async {
                if let url = URL(string: launchURL) {
                    UIApplication.shared.open(url, options: [:])
                }
            }
        }

        // Additional Data - THIS IS THE KEY PART for your custom data
        if let additionalData = event.notification.additionalData {
            print("OneSignal: Additional Data found: \(additionalData)")
            // Convert [AnyHashable: Any] to [String: Any] for proper JS serialization
            var convertedData: [String: Any] = [:]
            for (key, value) in additionalData {
                if let stringKey = key as? String {
                    convertedData[stringKey] = value
                }
            }
            clickData["additionalData"] = convertedData
            print("OneSignal: Converted additionalData: \(convertedData)")
        } else {
            print("OneSignal: No additionalData in notification")
        }

        // Notify JavaScript listeners
        if let plugin = self.plugin {
            print("OneSignal: Notifying JavaScript listeners")
            plugin.notifyListeners("notificationClicked", data: clickData, retainUntilConsumed: true)
        } else {
            print("OneSignal: WARNING - plugin reference is nil, cannot notify JavaScript")
        }
    }

    @objc public func setLogLevel(_ level: String) {
        let upper = level.uppercased()
        switch upper {
        case "VERBOSE":
            OneSignal.Debug.setLogLevel(.LL_VERBOSE)
        case "DEBUG":
            OneSignal.Debug.setLogLevel(.LL_DEBUG)
        case "INFO":
            OneSignal.Debug.setLogLevel(.LL_INFO)
        case "WARN", "WARNING":
            OneSignal.Debug.setLogLevel(.LL_WARN)
        case "ERROR":
            OneSignal.Debug.setLogLevel(.LL_ERROR)
        case "NONE":
            OneSignal.Debug.setLogLevel(.LL_NONE)
        default:
            print("Unknown log level: \(level), defaulting to INFO")
            OneSignal.Debug.setLogLevel(.LL_INFO)
        }
    }

    public func requestPermission(completion: @escaping (Bool) -> Void) {
        print("Requesting push notification permission")
        OneSignal.Notifications.requestPermission({ accepted in
            print("User accepted notifications: \(accepted)")
            completion(accepted)
        })
    }

    public func setExternalUserId(_ userID: String) {
        print("Setting external user ID to: \(userID)")
        OneSignal.login(userID)
    }

    public func clearExternalUserId() {
        print("Clearing external user ID")
        OneSignal.logout()
    }

    public func addTags(_ tags: [String: String], completion: @escaping (Bool) -> Void) {
        print("Adding tags: \(tags)")
        OneSignal.User.addTags(tags)
        completion(true)
    }

    public func removeTags(_ tags: [String], completion: @escaping (Bool) -> Void) {
        print("Removing tags: \(tags)")
        OneSignal.User.removeTags(tags)
        completion(true)
    }

    public func sendTag(_ key: String, _ value: String, completion: @escaping (Bool) -> Void) {
        print("Sending tag: \(key) = \(value)")
        OneSignal.User.addTags([key: value])
        completion(true)
    }

    public func removeTag(_ key: String, completion: @escaping (Bool) -> Void) {
        print("Removing tag: \(key)")
        OneSignal.User.removeTags([key])
        completion(true)
    }

}
